### ISSUE TITLE
Configurable concurrency level in reconciliation of shards

### ISSUE THREAD

**Issue by zcross** [Altinity/clickhouse-operator#1109]
Has anybody previously looked into modifying the worker implementation to support concurrent reconciliation of more than one shard at once? I have done a little searching in the project's issues and have only come across #376 which seems to be based on the same concern (time of operations on large clusters), though that's limited to creation and I'm interested in reconciliation in general.

Given sequential reconciliation over a cluster's shards, in which each statefulset pod takes O(single digit minutes) to complete, large clusters can take hours to reconcile. When the operation in question is some action needed to restore stability, that translates to long time to resolution (and in turn, potentially long periods of sustained negative impact).

My few concrete ideas for now:
- Orchestrate the sequential work in [this for loop over shards](https://github.com/Altinity/clickhouse-operator/blob/master/pkg/apis/clickhouse.altinity.com/v1/type_chi.go#L419) with a configurably-sized worker pool (this is extremely reductive and probably overlooking lots of other internal state)
- Add simple synchronization to the underlying map behind the `Registry` type (which I *think* is used within the scope of a single worker) – as well as all other internal mutable state used during shard reconciliation
- To avoid opting all current users of the operator into this behavior, make it configurable and use a default matching the status quo (n=1 concurrency, i.e., no additional concurrency)
- Use the fact that the shard layout is static to support either absolute number specification of concurrency, or a percentage specification. Both should have semantics similar to that of k8s' `MaxUnavailable` option of rolling updates (which I realize is not applicable to the stateful sets modeling each replica of each shard), i.e., phrasing how many shards will be affected in parallel (even though we should try to maintain the availability of any shards that use replication by reconciling a replica/host at a time).
- While this may not match every usage pattern of clickhouse, it should be a useful opt-in behavior for some use cases. 
- I'm not proposing adding intra-shard concurrency (e.g., I think we want to maintain the availability of at least one replica within each shard at all times, in configurations using replication).

I'm beginning to read the sources now, and ramping up on the development workflow for this project, but figured I'd ask first for a gut check on both the feasibility of and the appetite for such a change. I'm hoping it's not too controversial (e.g., https://github.com/Altinity/clickhouse-operator/issues/376#issuecomment-627812696 suggests that at least one other person thinks it's worth consideration). If it is feasible and sounds like a reasonable change (at least in the abstract, before you have a PR to review), I'm happy to take a shot at implementing the change (not just filing an issue). Of course, I'm also not at all against a maintainer leading the charge since I have to ramp up on the source, the dev workflow, and the contributing guidelines (all well documented) of this project, not to mention that this is my first foray into the implementation of a k8s operator.

Thanks for reading!

**Comment by sunsingerus** [comment]
Concurrent reconciliation has been discussed for a while. If we'll have a green light for this change, we need to discuss (at least) an architecture before starting implementation. We may also met some gotchas in "not-that-obvious" parts, such as concurrent status update. I'd rather invest some efforts in bringing operator ready for "concurrent reconciliation" (which it is not ready yet in some ares) before diving into it. But this all is a matter of discussion, of course.

**Comment by zcross** [comment]
Thanks for the reply! I'm definitely operating at a surface level here, so I think I am totally unaware of most of the gotchas. I'm more or less approaching this with a mental model along the lines of "if the work done during shard reconciliation is idempotent and always performed on a single statefulset pod/host at a time, then maybe we can perform a few of those operations in parallel even if some internal bookkeeping requires synchronization."

Of course, there may be incorrect assumptions built into that, such as the part about the operations being very cleanly isolated to single hosts and their CMs/PVCs/etc. 

Anyway, I was curious about how a naive implementation would work (at all) locally so I've been doing some testing with minikube (bringing up and modifying clusters of a few shards and watching the operator logs during reconciliation). The code is in no state near ready for a PR (not limited to the gotchas you've alluded to, or the fact that the new behavior is not configurable as I've expressed a desire for). [However, my sketch can be found here if you're interested.](https://github.com/Altinity/clickhouse-operator/compare/0.20.3...zcross:clickhouse-operator:zcross/concurrent-shard-reconciliation?expand=1)

By the way, I'm not totally following you here, could you clarify a little more?
> I'd rather invest some efforts in bringing operator ready for "concurrent reconciliation" (which it is not ready yet in some ares) before diving into it.

**Comment by zcross** [comment]
Some thoughts about the strategy for orchestrating the new parallel work:

First (not new, just making explicit): One cluster at a time (no concurrency change) – this means that in a multi-cluster configuration, problems resulting from things like misconfiguration will be detected early and limited to the first cluster

Second (new): Run reconciliation logic for the first shard (and its hosts) before fanning out – this both ensures that some of the "common" (e.g., configmap) reconciliation is successfully complete first, but also provides the same "fail early, with limited impact" property of the above. It ensures that _at least_ the first shard and all of its hosts are reconciled successfully before beginning the concurrent reconciliation of further shards, hopefully limiting damage in circumstances like a misconfiguration that is guaranteed to fail.

Third (not new, making explicit again): Strictly work in serial within shards, as is already the case, maintaining shard availability in cluster configurations with a replication factor > 1.

From a performance standpoint, the most significant contributor to overall reconciliation time changes from being linear in clusters*shards*replicas to linear in cluster*replicas (note: I'm leaving out references to some of the unchanged cluster-level pre/post shard operations). For instance, if reconciliation takes about 4min/replica, an RF=2 cluster of 100 shards would (*naively*) take (1 initial shard + 1 parallel remaining shards) * (2 replicas/shard) * (4 min/replica) = 16min instead of (100 serial shards) * (2 replicas/shard) * (4 min/replica) = 800 min. That concrete example is extremely naive, but the claim is just that we are replacing one linear factor (shards) with "a constant" (or at least some major reduction of that factor). Of course, that's idealized and ignoring the fact that my hypothetical approach takes a configurable max concurrency level (and shard availability percent) rather than always spinning up a goroutine per shard and doing them all at once.

Playing with a sketch implementation I have in minikube (one that will probably fail PR review in its current form), I am seeing that performance benefit mostly realized, even if not matching the imprecise napkin math exactly. I think I'll try to clean it up, see how it handles the end to end test suite, and – only if those two steps are successful – try to continue this conversation within a pull request.

Thanks!

**Comment by sunsingerus** [comment]
Talking about concurrent run, there are at least two items (these are what I can name out of my mind right away, there may be something else as well) that needs to be taken care of in terms of concurrent access: 
- registry of reconciled objects 
- CHI status. 
They are not concurrent-safe ATM.
Also not sure about announcer, because it can not only write logs, but also emit events and yet again, update CHI status.

**Comment by sunsingerus** [comment]
I would rather implement it with small steps, starting with concurrent-unsafe objects preparation and adding concurrency at the very end

**Comment by zcross** [comment]
Agreed, it's likely to be a large-ish PR if it's all done in branch. I think my "sketch" branch ventured into adding simple mutex based synchronization into the registry type, but I haven't yet done the same for status (and I'm slowly learning by reading how the status is updated, and other things like the various `MergeFrom` calls on the `chi` type). I appreciate the input on the landscape of "things needing synchronization;" those two were on my radar but I wasn't certain if that was the complete set of mutable state (I continue to learn how the operator is designed as I iterate and branch out from the source files I initially focused on).

**Comment by zcross** [comment]
Is it fair to characterize the internal mutable state as bookkeeping that can be lost at any point in time (e.g., including operator restarts)? I'm under the impression that all operations are planned based on introspecting the environment itself prior to the operation (e.g., determining a create vs an update of a CM begins by querying for the CM), even if things like the registry/the chi struct provide some surface for metadata bookkeeping. Just looking to validate my mental model as I proceed with your recommended approach.

**Comment by zcross** [comment]
I *think* the three PR's I have opened capture all of the original thoughts on this issue (but are still awaiting review/input) and I wouldn't rule out additional PRs if the maintainers think of other surface areas that need tweaks. Thanks for reading!

**Comment by zcross** [comment]
Note to self / people following: I think I'll need to think a little about the underlying k8s client call volume increase that utilizing this feature produces; I have gone from testing in minikube to testing in a "real" k8s environment to get a better feel for performance and have seen some nontrivial amount of client side throttling, e.g.,:
```
request.go:668] Waited for 3.196177151s due to client-side throttling, not priority and fairness, request: GET:https://[redacted]/api/v1/namespaces/zachtest/pods/chi-mychi-data-a-6-1-0
```

**Comment by zcross** [comment]
Hi @alex-zaitsev, no urgency here but thought I'd check in to see if you all had the review capacity to take a look at this body of work sometime. I know you all were busy with the `0.21.0` release. I have some rebasing/cleanup to do that is self-obvious but otherwise in a holding pattern waiting further maintainer review. Thanks!

**Comment by zcross** [comment]
For any interested readers, the current status of this work is that the suggested preliminary PRs that added some synchronization to previously concurrency-unsafe objects are landed:
- #1115 
- #1119 

The final PR that adds the actual concurrent reconciliation is open:
- #1124

... but I admit that the lack of e2e test coverage is not ideal, even if the feature ships as inert/off-by-default and requires opting in. I have had a pretty hard time running the dev environment and particularly the e2e tests as a Docker newbie limited to arm64 devices (e.g,. Mac M1/M2s). I'd at least like n>0 test cases that enable the concurrent reconcile feature and otherwise resemble existing test cases for exercising reconciliation of various shard layouts, as well as reconciliation failure modes and the associated recovery that is expected.

Thanks for all the review and input so far!

**Comment by zcross** [comment]
Marking this closed as @sunsingerus merged into `0.21.1`. Thanks all!

**Comment by zcross** [comment]
Just validated a "real" usage of this on a big-ish cluster that would usually take hours and it was reconciled in tens of minutes. Thanks again! 😀

**Comment by sunsingerus** [comment]
The first item needs additional research, as to the second and third - default operator's behavior has changed to the following:
